### DISCUSSION TITLE
Use custom OAuth2DeviceUserCodeProvider instead of default for Oauth2 Device Grant flow

### DISCUSSION THREAD

**Discussion by mikethebeer** [keycloak/keycloak#20815] [Q&A] [answered]
I'm trying to replace the `DefaultOAuth2DeviceUserCodeProvider` with a custom implementation that should be used in the OAuth2 Device Grant flow. The custom provider just reduces the amount of chars generated as device code.
However, I'm unable to replace it.

Things I did:

* Implement the custom Provider and its factory

```kotlin
class CustomOauth2DeviceCodeProvider implements OAuth2DeviceUserCodeProvider {
  ...
    override fun getId(): String {
        return "op-oauth2"
    }
}
```

* create a new service provider configuration with name: `org.keycloak.models.OAuth2DeviceUserCodeProviderFactory` that contains the Factory name of the custom user provider.

Custom provider gets loaded on startup:

```
keycloak_1  | 2023-06-06T12:58:31.026568608Z 2023-06-06 12:58:28,560 WARN  [org.keycloak.services] (build-41) KC-SERVICES0047: op-oauth2 (com.example.CustomOAuth2DeviceCodeProviderFactory) is implementing the internal SPI oauth2DeviceUserCode. This SPI is internal and may change without notice
```

As seen in the Admin UI it gets loaded (however, at 2nd position):

![grafik](https://github.com/keycloak/keycloak/assets/6863216/cefb28b3-92ac-42c8-94d3-ef707c185272)

The first device authentication request still returns the default 2x4 alphanumeric chars as device code

```
curl --location --request POST 'http://localhost:8081/realms/stg/protocol/openid-connect/auth/device' \
--header 'Content-Type: application/x-www-form-urlencoded' \
--data-urlencode 'client_id=myclient' \
--data-urlencode 'scope=offline_access openid'
``` 

```
..."user_code":"GOXL-MXLQ",...
```

:cry: 

Now, I thought about disabling the `default` provider by using the CLI

```
kc.sh start --spi-oauth2DeviceUserCode-default-enabled=false"
```

and still the same result appears. Also, the `default`  provider still gets loaded (as seen in the Admin UI).

What did I do wrong?

**Comment by adamkobor** [comment, marked as answer]
we already discussed this with @mikethebeer in person, but it could be useful for others too: the trick was to override the `order` method of `ProviderFactory` (providing a **higher** number makes your custom provider override the built-in one)
